tools/azure-pipelines-tasks-json-generate.mjs

```diff
@@ -0,0 +1,59 @@
+import os from 'os';
+import { promisify } from 'util';
+import fs from 'fs-extra';
+import g from 'glob';
+import JSON5 from 'json5';
+import shell from 'shelljs';
+import Git from 'simple-git';
+import path from 'upath';
+
+const glob = promisify(g);
+
+const localPath = path.join(os.tmpdir(), 'azure-pipelines-tasks');
+
+await (async () => {
+  await fs.ensureDir(localPath);
+  const git = Git(localPath);
+
+  if (await git.checkIsRepo()) {
+    await git.pull();
+  } else {
+    await git.clone(
+      'https://github.com/microsoft/azure-pipelines-tasks.git',
+      '.'
+    );
```

Comment:
Are you sure that a full clone is necessary and no other solution such as blobless cloning suffices?

Reply:
It's pretty big:

```
❯ git clone https://github.com/microsoft/azure-pipelines-tasks.git
Cloning into 'azure-pipelines-tasks'...
remote: Enumerating objects: 270480, done.
remote: Total 270480 (delta 0), reused 0 (delta 0), pack-reused 270480
Receiving objects: 100% (270480/270480), 482.08 MiB | 6.81 MiB/s, done.
Resolving deltas: 100% (160273/160273), done.
```

Reply:
I experimented with a couple of optimized clones. Here's the current full clone from scratch

```
$ time node tools/azure-pipelines-tasks-json-generate.mjs
...
________________________________________________________
Executed in  122.07 secs    fish           external
   usr time   94.82 secs  264.00 micros   94.82 secs
   sys time   48.89 secs  652.00 micros   48.89 secs
```

I experimented with `--filter=blob:none`, `--filter=tree:0`, and `sparse-checkout`. `--filter=blob:none` and `--filter=tree:0` result in quicker initial clones, but massively slow down the overall script due to doing thousands of tiny requests on demand. I don't have timing data for these as I killed them after 2 minutes.

Here's the `sparse-checkout`:

```
$ time node tools/azure-pipelines-tasks-json-generate.mjs
________________________________________________________
Executed in  116.29 secs    fish           external
   usr time   98.13 secs  182.00 micros   98.13 secs
   sys time   45.41 secs  456.00 micros   45.41 secs
```

IMO it's not worth the additional complexity of sparse clones for what is within the margin of error.

Reply:
Unfortunately we do need the entire history to find all the historical versions.